### PR TITLE
Search for the prerequisites from package configuration file.

### PR DESCRIPTION
**Pull request by blattms**
That is the way it is done in other modules, too. We need this
at least for libecl as this target is used in opm-common_LIBRARIES
and not defined if there was no find_package(ecl) call. That call
is now made during execution of find_package_deps(opm-common).

This is an alternative to #457 and prevents users from manually searching for libecl.

### CODE REVIEW

File: opm-common-prereqs.cmake [hunk 1]
```diff
@@ -25,3 +25,5 @@ else()
         "Boost 1.44.0
           COMPONENTS system unit_test_framework REQUIRED")
 endif()
+# We the a defined target for libecl
```
Comment (akva2) [inline opm-common-prereqs.cmake]: wat?

Comment (blattms) [reply opm-common-prereqs.cmake]: thanks. Improved

**Comment by akva2** [comment]
i prefer this but then we can remove the other lib handling for ecl in OpmFind, no ?

**Comment by blattms** [comment]
Will give that a try and report back.

**Comment by blattms** [comment]
@akva2 you mean all of [OpmFind.cmake#L165-L173](/OPM/opm-common/blob/19afcbf769e8c4e11316248be49156ea2dd4b43d/cmake/Modules/OpmFind.cmake#L165-L173)?

**Comment by akva2** [comment]
i mean the lib part. some stuff still might want the header paths.

**Comment by blattms** [comment]
to the very least we need to have `set(ECL_LIBRARIES ecl)` as libecl does not export the target name of the library.

**Comment by akva2** [comment]
yes that is what i meant. library as ecl, only dig out headers in case something wants those.

**Comment by blattms** [comment]
Seems like we still need these. If I just use `set(ecl_LIBRARIES ecl)` then opm-common does not compile:
```
[ 19%] Building CXX object CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o
/usr/bin/c++   -DHAVE_CONFIG_H=1 -I/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel -I/home/mblatt/src/dune/opm-2.6/opm-common/external/cjson -I/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel/include -I/home/mblatt/src/dune/opm-2.6/opm-common  -std=c++14 -pipe -Wall -Wno-unknown-pragmas -pthread -O3 -mtune=native -DNDEBUG -fPIC   -std=c++14 -o CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o -c /home/mblatt/src/dune/opm-2.6/opm-common/src/opm/parser/eclipse/Deck/Deck.cpp
In file included from /home/mblatt/src/dune/opm-2.6/opm-common/opm/parser/eclipse/Deck/Deck.hpp:32:0,
                 from /home/mblatt/src/dune/opm-2.6/opm-common/src/opm/parser/eclipse/Deck/Deck.cpp:23:
/home/mblatt/src/dune/opm-2.6/opm-common/opm/parser/eclipse/Units/UnitSystem.hpp:28:30: fatal error: ert/ecl/ecl_util.h: No such file or directory
 #include <ert/ecl/ecl_util.h>
                              ^
compilation terminated.
CMakeFiles/opmcommon.dir/build.make:873: recipe for target 'CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o' failed
make[2]: *** [CMakeFiles/opmcommon.dir/src/opm/parser/eclipse/Deck/Deck.cpp.o] Error 1
make[2]: Leaving directory '/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel'
CMakeFiles/Makefile2:1321: recipe for target 'CMakeFiles/opmcommon.dir/all' failed
make[1]: *** [CMakeFiles/opmcommon.dir/all] Error 2
make[1]: Leaving directory '/home/mblatt/src/dune/opm-2.6/opm-common/opm-parallel'
Makefile:138: recipe for target 'all' failed
make: *** [all] Error 2
```

**Comment by blattms** [comment]
Nevermind my last comment. I had a typo in the variable name ecl_LIBRARIES. will retest

**Comment by akva2** [comment]
nice.

File: opm-common-prereqs.cmake [hunk 2]
```diff
@@ -25,3 +25,7 @@ else()
         "Boost 1.44.0
           COMPONENTS system unit_test_framework REQUIRED")
 endif()
+# We need a defined target for libecl when linking.
+# The definition of the target is done implicitly below when
+# libecl is searched for.
+find_package_deps(opm-common)
```
Comment (bska) [inline opm-common-prereqs.cmake]: Do you really mean `opm-common` here (in `opm-common-prereqs.cmake`)?  Is `opm-common` a prerequisite for itself?

Comment (akva2) [reply opm-common-prereqs.cmake]: it's the prefix for the variables used inside the macros. opm-common_CONFIG_VAR, opm-common_DEPS.

Comment (bska) [reply opm-common-prereqs.cmake]: > it's the prefix for the variables used inside the macros

Ah, right.  Thanks for clarifying and sorry for the confusion.

**Comment by blattms** [comment]
Fixed one commit message with a forced push. Sorry.

I have tested this branch (cherry-picked to releases/2018.04) with @joakim-hove's described in #456 and it worked on my system with extra patches to sunbeam. Additionally I tried it together with opm-grid and dumux. In contrast to @andlaus I can build dumux without problems (by passing -DUSE_MPI=ON to cmake to activate MPI for OPM). Either his setup is very different (which I cannot judge because his report is too detailed and might well be) or I fixed some bugs in this branch without really noticing.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
BTW we are still only check ENABLE_ECL_INPUT and neglect ENABLE_ECL_OUTPUT in the package configuration file. Is that really correct. i.e. does ENABLE_ECL_OUTPUT being ON imply the same for ENABLE_ECL_OUTPUT?

**Comment by akva2** [comment]
OUTPUT implies INPUT, https://github.com/OPM/opm-common/blob/master/CMakeLists.txt#L12

both are exported through the config file, https://github.com/OPM/opm-common/blob/master/CMakeLists.txt#L73

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by akva2** [comment]
this will still break due to bugs in the boost find module.

**Comment by akva2** [comment]
might have found a local issue on jenkins. let's try again and see if it fixed anything.

**Comment by akva2** [comment]
jenkins build this please

**Comment by blattms** [comment]
When I looked at the output, I saw a linking error in upscaling (boost iostream). Will test this locally this afternoon.

**Comment by akva2** [comment]
i cannot trigger it locally, only on jenkins, which is why i'm looking into local issues.

**Comment by blattms** [comment]
I have to explicitly use `DINSTALL_BENCHMARKS=ON` with cmake to reproduce the problem on my machine. Otherwise upscale_relperm_benchmark will not be built.

**Comment by blattms** [comment]
`-lboost_iostream` seems to missing from the linker line. I guess that problem is in opm-upscaling?

**Comment by blattms** [comment]
Well, somehow searching for the same lib multiple times must have been lost. Ergo we neglect the components requested in opm-upscaling. Will fix.

**Comment by blattms** [comment]
jenkins build this please

**Comment by blattms** [comment]
This should be fixed now. 

Indeed the reason for this failure was that we only searched for Boost once after my changes. The reason was that I stripped the variable opm_common_processed that basically made us search for everything required above of the line for opm-common in opm-module-prereq.cmake multiple times (including Boost by accident or by design). With my changes these multiple searches were removed. With my last commit another search is now always triggered if the argument list to find_package contains the word COMPONENTS.

**Pull request merged**